homeassistant/components/bluesound/media_player.py

```diff
@@ -829,83 +812,118 @@ def is_master(self):
 
         return attributes
 
     def rebuild_bluesound_group(self):
         """Rebuild the list of entities in speaker group."""
-        if self._group_name is None:
+
+        if not self.is_master:
+            return None
+
+        # players not initialized this method will be triggered later
+        if self.hass is None:
             return None
 
-        bluesound_group = []
+        slaves = []
+        slave_node = self._sync_status.get("slave")
+        nodes = slave_node if isinstance(slave_node, list) else [slave_node]
 
-        device_group = self._group_name.split("+")
+        for slave in nodes:
+            slave_id = slave.get("@id")
+            slave_port = slave.get("@port")
+            slaves.append(f"{slave_id}:{slave_port}")
 
-        sorted_entities = sorted(
-            self._hass.data[DATA_BLUESOUND],
-            key=lambda entity: entity.is_master,
-            reverse=True,
+        bluesound_group: list[BluesoundPlayer] = []
+        bluesound_group.append(self)
+
+        bluesound_group.extend(
+            device for device in self.hass.data[DATA_BLUESOUND] if device.id in slaves
         )
-        bluesound_group = [
-            entity.name
-            for entity in sorted_entities
-            if entity.bluesound_device_name in device_group
-        ]
 
         return bluesound_group
 
-    async def async_unjoin(self):
-        """Unjoin the player from a group."""
-        if self._master is None:
-            return
+    async def async_join_players(self, group_members: list[str]) -> None:
+        """Join `group_members` as a player group with the current player."""
+
+        devices = [
+            device
+            for device in self.hass.data[DATA_BLUESOUND]
+            if device.entity_id in group_members
+        ]
+
+        master: BluesoundPlayer
+        if self.is_master:
+            master = self
+        elif self.is_grouped:
+            master = self.master
+        else:
+            # try to be smart and check if any other devices is already in group
+            other_masters = [device for device in devices if device.is_master]
+            if len(other_masters) > 0:
+                master = other_masters[0]
+                devices.remove(master)
+                devices.append(self)
+            else:
+                other_slaves = [device for device in devices if device.is_grouped]
+                if len(other_slaves) > 0:
+                    master = other_slaves[0].master
+                    devices.append(self)
+                    devices.remove(other_slaves[0])
+                    _LOGGER.warning("5 %s", devices)
+                else:
+                    master = self
 
-        _LOGGER.debug("Trying to unjoin player: %s", self.id)
-        await self._master.async_remove_slave(self)
+        await master.add_slaves(devices)
 
-    async def async_add_slave(self, slave_device):
+    async def add_slaves(self, slave_devices):
```

Comment:
I'd recommend to rename the library methods to something more modern, instead of slave/master.

Reply:
slave/master is the same terminology that bluesound is using in their API. I think the way to go is to match their terms. But if you want me to change it I will of course do so.

Reply:
It's just a recommendation. If it makes sense or not you need to decide who knows the device.